Centralize Xcode version definition in Buildkite pipeline

mokagio opened the pull request:
See discussion at https://github.com/wordpress-mobile/WordPress-iOS/pull/16787#discussion_r676240991

If [Buildkite runs fine](https://buildkite.com/wordpress-mobile/wordpress-ios/builds/1674#755fd55d-af8b-4f85-9811-b9af6edc023a), then we're good to go.

I verified it locally by parsing the YML to see if all the anchors were resolved correctly.

## Regression Notes
1. Potential unintended areas of impact
N.A.


2. What I did to test those areas of impact (or what existing automated tests I relied on)
N.A.

3. What automated tests I added (or what prevented me from doing so)
N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

mokagio committed:
Centralize Xcode version definition in Buildkite pipeline

mokagio committed:
DRY the entire step environment definition in the Buildkite pipeline

AliSoftware commented on .buildkite/pipeline.yml:
```diff
@@ -5,6 +5,17 @@ use_plugins: &use_plugins
           bucket: "a8c-repo-mirrors"
           repo: "wordpress-mobile/wordpress-ios/"
 
+xcode_version: &xcode_version
+  IMAGE_ID: xcode-12.5.1
+
+env: &env
+  # If you'll need more than one key-value, update to:
+  #
+  # env:
+  #   <<: *xcode_version
+  #   other_key: other_value
+  env: *xcode_version
+
```
1. In your comment to suggest how to handle more than one value, you forgot to set the anchor: `env: &env`
2. Overall I'm not sure it's worth extracting the `xcode-version:` (and its `IMAGE_ID` key) entry separately as the only place it's used is in the env…
So what about merging them already, allowing you to get rid of the comment altogether?

```suggestion
x-common-env: &env
  env:
    IMAGE_ID: xcode-12.5.1
```

PS: I use `x-common-env` as the key just to emphasis that the name of this key:

1. Is purely arbitrary, and doesn't have to be `env:` here, because what the `&env` YAML anchor captures is what's after `&env` (so the key is not captured by the anchor, which is why you have to specify the `env:` key after/inside `&env` if you want it to be outputed when you use `*env` to "paste" it later in the YAML)
2. Is not a key that has a specific meaning in the BuildKite specification – hence starting it with `x-` in an attempt to convey the fact that it's a custom key from us, not a BuildKite key, similar to the old (and now deprecated, but 🤷 ) convention that HTTP used for experimental/non-standard headers)

---

Another approach would be to not include the `env:` key inside the `&env` anchor, and instead repeat it at every call site explicitly. That kinda reduces the goal of the anchor DRY-ing stuff and avoiding repeating that `env:` key every time, but I think that it would actually look nicer in this context, since this anchor definition here would then look like this:

```suggestion
x-common-env: &env
  IMAGE_ID: xcode-12.5.1
```
And at call site like [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/16929/files#diff-a3991ebf1475eb82acab13946ffc1eca02b43917f6d5bec3898f45c8b0b9bd53L16-R27) instead of using the YAML `<<:` merge operator, it would instead look like this, which is fine imho:
```yml
command: ".buildkite/commands/build-for-testing.sh"
env: &env
<<: *use_plugins
```

In any case, both solutions work so I'll let you decide which one you prefer (it doesn't really matter to me I'm fine with both 😉), as long as it gets simplified as I don't think we need the `xcode_version` to be extracted as opposed to being directly defined inline in the `env` dictionary we DRY'd here.

mokagio committed:
Merge Xcode image and environment definition in Buildkite pipeline

See suggestion from @AliSoftware here:
https://github.com/wordpress-mobile/WordPress-iOS/pull/16929#discussion_r677323734

I liked the extracted Xcode version because it made it easier to find
it, IMHO, but I'll admit it was a bit over-engineered, in particular
given we don't have any need to use it on its own right now.

I chose the approach in this commit because I like how it forces each
step to explicitly define the `env` key. I feel doing so keeps the step
syntax more explicit. I our pipeline files grow in size, a reader won't
have to jump to the definition of the anchor to know it's used for the
`env` key, which they would have had to do with the `<<: *common_env`
approach.

I also decided to use underscores instead of the suggested dashes to
keep it in line with the other anchor already defined (`use_plugin`) and
the Buildkite spec (
https://github.com/buildkite/pipeline-schema/blob/f7a0894074d194bcf19eec5411fec0528f7f4180/schema.json
) which uses underscores, e.g. in `artifacts_paths`.

mokagio committed:
Add `x_` prefix to `use_plugin` YML anchor in Buildkite pipeline

For consistency with the other custom anchor, `x_common_env`.

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -1,10 +1,26 @@
-use_plugins: &use_plugins
+x_use_plugins: &use_plugins
```
Updated with the `x_` prefix for consistency with the one below.

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -1,10 +1,26 @@
+#   env:
+#     <<: *common_env
+#     other_key: "other value"
+x_common_env: &common_env
+  IMAGE_ID: xcode-12.5.1
```
I this approach in this commit because I like how it forces each step to explicitly define the `env` key. If our pipeline files grow in size, a reader won't have to jump to the definition of the anchor to know it's used for the `env` key, which they would have had to do with the `<<: *common_env` approach.


```yml
env: *common_env
# vs
<<: *common_env_with_env_root_in_it
```

I also decided to use underscores instead of the suggested dashes to keep it in line with the other anchor already defined (`use_plugin`) and the Buildkite [spec](buildkite/pipeline-schema@f7a0894/schema.json) which uses underscores, e.g. in `artifacts_paths`.

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -13,8 +29,7 @@ steps:
-    env:
-      IMAGE_ID: xcode-12.5.1
+    env: *common_env
     <<: *use_plugins
```
For consistency, we might want to update this, too, to use something like `plugins: *common_plugins`.

I want to hear what the reviewers have to say on my approach first, though.

AliSoftware replied:
Definitely.

See also [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3764#discussion_r678088237), and the full explanation/discussion (I went full-verbose-Olivier again sorry…) [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3758#discussion_r678072609) on @oguzkocer 's PR, where we wondered the same thing

mokagio commented:
Thank you for the review @AliSoftware . I addressed your suggestions and Buildkite is [still green](https://buildkite.com/wordpress-mobile/wordpress-ios/builds/1686).

AliSoftware commented on .buildkite/pipeline.yml:
```diff
@@ -1,10 +1,26 @@
-use_plugins: &use_plugins
+x_use_plugins: &use_plugins
     plugins:
       - automattic/bash-cache#v1.0.0: ~
       - automattic/git-s3-cache#add/vm-host-support:
           bucket: "a8c-repo-mirrors"
           repo: "wordpress-mobile/wordpress-ios/"
 
+# Common environment values for the `env` key.
+#
+# To add more, simply add another entry at the same indentation as `IMAGE_ID`.
+#
+# Usage in a step with only these values:
+#
+#   env: *common_env
+#
+# Usage in a step with other values:
+#
+#   env:
+#     <<: *common_env
+#     other_key: "other value"
+x_common_env: &common_env
+  IMAGE_ID: xcode-12.5.1
+
```
What I don't really like about this approach of using a key before the anchor that has the exact same name as the anchor, is that it gives the incorrect impression, especially to someone new to YAML anchors, that this is a requirement and that an anchor always have to be defined on a key of the same name. It hasn't.

In fact the only reason why we usually put an arbitrary key like that before the definition of the anchor is to nicely isolate the node we want to define an anchor on, and to not define that anchor at the top level of the YAML, but instead "namespace" that definition to avoid polluting the rest of the YAML.

So to clarify all that and imho make it more readable, and because the only role is to define anchors at the top of the YAML but still not at top-level in the root node, I personally prefer using an array for the namespacing, with one item for each anchor to define. That also avoids the confusion for people new to anchors.

```yaml
# Common YAML nodes we will reuse multiple times in the YAML
common_params:
   # Common plugins to use with the `plugins:` key
 - &common_plugins
   - automattic/bash-cache#v1.0.0: ~
   - automattic/git-s3-cache#add/vm-host-support:
       bucket: "a8c-repo-mirrors"
       repo: "wordpress-mobile/wordpress-ios/"

  # Common environment values for the `env` key.
- &common_env
  IMAGE_ID: xcode-12.5.1
```

mokagio replied:
I really, really like this. In particular, it's clean and easy to extend: one doesn't need extra documentation comments or policies to explain how to add new shared values.

oguzkocer replied:
> What I don't really like about this approach of using a key before the anchor that has the exact same name as the anchor, is that it gives the incorrect impression, especially to someone new to YAML anchors, that this is a requirement and that an anchor always have to be defined on a key of the same name.

😅

jkmassel commented:
This YAML stuff is totally over my head, but I love what this PR is aiming to accomplish and I'll let you two figure it out 😅

mokagio commented:
@jkmassel 

> This YAML stuff is totally over my head

In case it wasn't apparent, I'm totally winging it 🤣 So don't feel alone in that.

mokagio committed:
Update Buildkite common environment definition to remove root key

Hat tip @AliSoftware.

See discussion here:
https://github.com/wordpress-mobile/WordPress-iOS/pull/16929#discussion_r678112821

mokagio committed:
Move Buildkite common plugins definition in common_params

This keeps the pipeline configuration consistent.

See previous commit and discussion here:
https://github.com/wordpress-mobile/WordPress-iOS/pull/16929#discussion_r678112821

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -1,9 +1,14 @@
-          bucket: "a8c-repo-mirrors"
-          repo: "wordpress-mobile/wordpress-ios/"
+# Nodes with values to reuse in the pipeline.
+common_params:
```
Alternative syntaxes I tried here:

```yml
x_common_params:

<<:

_:
```

The all worked. I though of the one with the `x_` prefix because of the `x` prefix convention mentioned earlier in the discussion. I thought of `<<` and `_` to avoid having a key and tell the reader that this node was something we defined outside of the pipeline spec.

I opted not to use either `<<` or `_` because they are not that clear; if we removed the comment, the purpose of the syntax might get lost. Conversely, `common_params` and `x_common_params` are self-explanatory even without the comment.

I'm happy to use the `x_` prefix, but I wonder if we really need the extra information it carries in this configuration.

AliSoftware replied:
I'm ok with not having the x prefix.

One thing that could be nice, despite this key being totally arbitrary, is to try to always use the same name for it across the various YAML and repositories, just for consistency. So we might want to bikeshed it a bit (outside of this PR, eg in Slack) to find one that everyone is happy with. But that's really nitpicks so nbd.

mokagio commented:
@AliSoftware addressed your latest, and very nice, suggestion. The new syntax [parses correctly on Buildkite](https://buildkite.com/wordpress-mobile/wordpress-ios/builds/1715).

Mind having a new look? Thanks.

AliSoftware reviewed (approved):
✨

jkmassel reviewed (approved):
Love how clean this is!!

Merging it in order to rebase #16935
